Energy postprocessors

casperversteeg opened the pull request:
Added some postprocessors to monitor total kinetic and strain energy, completing the triage with the FractureEnergy postprocessor.
Also includes tests to verify they work--which they do. :)

hugary1995 commented:
I think we don't need the StrainEnergy postprocessor. There is a Material called ElasticEnergyDensity, you can use it to compute _elastic energy density_ and then use a postprocessor to integrate it.

hugary1995 commented:
All tests you've added are good. The only one that's a little bit weird is the kinetic energy one where you don't actually have any inertia -- but it is fine.

I think we need a bunch of other tests:
- KE+SE, with newmark beta
- KE+SE, with central difference
- KE+SE+FE, where FE is fracture energy.

Ideally we would want KE, SE, FE to be on the same order of magnitude and verify that KE+SE+FE ~= WORK.
I've done this before so let me know if you need help.

casperversteeg commented:
I added the tensile problem as a benchmark, and kept your tests. I modified the headers of the tests I wrote to make it more clear what they're doing. 
I also kept the StrainEnergy postprocessor. Even though there is an elastic energy postprocessor, it doesn't duplicate a whole lot of code, and helps reduce input file bloating. 
I also added a postprocessor that returns the number of fixed point iterations required per time-step. To get it to work I added a method `getSolveObj()` to the `FixedPointTransient` executioner. Didn't really see a cleaner way of doing that.

hugary1995 commented on src/postprocessors/FPIterations.C:
```diff
@@ -0,0 +1,72 @@
+void
+FPIterations::timestepSetup()
+{
+  if (_fp_problem->time() != _time)
```
This comparison is dangerous. Instead of keeping track of the number of FPIterations in this postprocessor, we should simply provide an interface in the FixedPointSolve object to return the number of iterations. This postprocessor should just make use of that API.

casperversteeg replied:
I don't quite understand why this is dangerous per se. The `NumNonlinearIterations` postprocessor does a similar thing. The `_time` parameter also get initialized to `-numeric_limits<Real>::max()`, so this will ostensibly only get executed on the first time step. The times I've used it as is hasn't given me any issues, but I'm curious to hear your thoughts on what an interface in `FixedPointSolve` would look like.

hugary1995 commented on src/postprocessors/FPIterations.C:
```diff
@@ -0,0 +1,72 @@
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
```
header

hugary1995 commented on src/postprocessors/KineticEnergy.C:
```diff
@@ -0,0 +1,41 @@
+KineticEnergy::validParams()
+{
+  InputParameters params = ElementIntegralPostprocessor::validParams();
+  params.addClassDescription("computes the total fracture energy of the form $\\int_\\body "
```
Wrong description

casperversteeg replied:
Fixed

hugary1995 commented on src/postprocessors/KineticEnergy.C:
```diff
@@ -0,0 +1,41 @@
+    _vel_var(_ndisp)
+{
+  for (unsigned int i = 0; i < _ndisp; ++i)
+    _vel_var[i] = &adCoupledDot("displacements", i);
```
use coupledDot instead of adCoupledDot

hugary1995 replied:
Actually this postprocessor should be templated by is_ad. Take a look at moose/modules/tensor_mechanics/src/postprocessors/MaterialTensorIntegral.C for example.

hugary1995 commented on src/postprocessors/KineticEnergy.C:
```diff
@@ -0,0 +1,41 @@
+Real
+KineticEnergy::computeQpIntegral()
+{
+  Real ans = 0;
```
There is an easier (and better in terms of stacks created) way of writing this:
```
RealVectorValue v((*_vel_var[0])[_qp], (*_vel_var[1])[_qp], (*_vel_var[2])[_qp]);
return 0.5 * _rho[_qp] * v * v;
```

casperversteeg replied:
Unfortunately, since (*_vel_var[2]) may not be given in a 2D problem, this will result in a segmentation fault. So we can either make a triplet of variables in the declaration of the class, i.e. 
```
_u( _ndisp >= 1 ? coupledDot("displacements", 0) : 0.0),
_v( _ndisp >= 2 ? coupledDot("displacements", 1) : 0.0),
_w( _ndisp == 3 ? coupledDot("displacements", 2) : 0.0) 
```
And then create the vector, or do it with the for-loop method.

hugary1995 commented on src/postprocessors/StrainEnergy.C:
```diff
@@ -0,0 +1,31 @@
+StrainEnergy::StrainEnergy(const InputParameters & parameters)
+  : ElementIntegralPostprocessor(parameters),
+    _base_name(isParamValid("base_name") ? getParam<std::string>("base_name") + "_" : ""),
+    _stress(getADMaterialPropertyByName<RankTwoTensor>(_base_name + "stress")),
```
This postprocessor should be is_ad templated as well. In some cases we will have AD stress and strain, sometimes we will have regular stress and strain. We really need both a StrainEnergy postprocessor and an ADStrainEnergy postprocessor.

casperversteeg replied:
Done.

hugary1995 commented on test/tests/postprocessors/KineticEnergy/totalKE.i:
```diff
@@ -0,0 +1,98 @@
+  []
+[]
+
+[Executioner]
```
This is a nice test! I guess I would like to see this in action using the NewmarkBeta integrator and the CentralDifference integrator.

casperversteeg replied:
Since this is not a hyperbolic problem, I _could_ make a separate 2D problem that uses time-integration, it just won't work with this one.

casperversteeg commented:
I put a fairly good dent in making as many things both AD and non-AD compatible as I thought made sense. Just took care of the easy ones. 
You'll notice that the LocalDissipation and Degradation materials are pretty much AD only, even though they're templated. I left them this way because I noticed you call `registerADMooseObject` in the source file, so in case your intent was to have them AD only, this doesn't change anything. If, however, you want to switch some of them over in the future (not sure why you would though...) this makes it as easy as un-commenting some lines. It also ensures that any classes inheriting from them can only be AD.

casperversteeg commented:
One thing these updates will require, is that input files must now use `ADGenericConstMaterial` and `ADFractureMaterial` everywhere, since the material values will be read as AD by the consuming kernels. It'll be a little bit of a pain to switch over input files, but it'll make it a lot easier for users in the future to figure out if they have to `get` a material property as regular or as AD, because they can just look at the specific material block they're using.
